File: fru1tworld/week26~30/week28/leet390/leet390.cpp
class Solution {
    public:
        int lastRemaining(int n) {
            int head=1;
            int step=1; 
            bool direction=true;
            while(n>1){
                if(direction) head += step; 
                else head += (n&1) == 0 ? 0 : step; 
                step*=2;
                n/=2;
                direction^=true;
            }
    
            return head;
        }
    };

File: iam454/week28/leet390.js
/**
 * @param {number} n
 * @return {number}
 */
var lastRemaining = function (n) {
  let head = 1;
  let step = 1;
  let leftToRight = true;

  while (n > 1) {
    if (leftToRight || n % 2 === 1) {
      head += step;
    }
    n = Math.floor(n / 2);
    step *= 2;
    leftToRight = !leftToRight;
  }

  return head;
};
